shpc/client/install.py

```diff
@@ -23,17 +24,126 @@ def main(args, parser, extra, subparser):
     # Update config settings on the fly
     cli.settings.update_params(args.config_params)
 
-    # And do the install
-    cli.install(
-        args.install_recipe,
-        force=args.force,
-        container_image=args.container_image,
-        keep_path=args.keep_path,
-    )
-    if cli.settings.default_view and not args.no_view:
-        cli.view_install(
-            cli.settings.default_view,
+    # Get the list of the user's installed software
+    installed_software = cli.list(return_modules=True)
+
+    # Upgrade a specific installed software
+    if args.upgrade:
+        # Check if the user specified a version
+        if ":" in args.install_recipe:
+            logger.exit(
+                "Please do not include the software version when using --upgrade argument."
+            )
+        # Check if the specific software is installed
+        if args.install_recipe not in installed_software:
+            logger.exit(
+                f"You cannot carry out an upgrade on {args.install_recipe} because you do not have it installed.\nInstall it first before attempting an upgrade.",
+                0,
+            )
+
+        # Does the user just want a dry-run?
+        if args.dryrun:
+            version_info = upgrade(
+                args.install_recipe, cli, args, dryrun=True
+            )  # This returns the latest version if its available, else returns None
+            if version_info:
+                logger.info(
+                    f"You do not have the latest version installed.\nLatest version avaiable is {version_info}"
+                )
+            else:
+                logger.info(
+                    f"You have the latest version of {args.install_recipe} installed."
+                )
+
+        # Upgade the software
+        else:
+            upgrade(
+                args.install_recipe,
+                cli,
+                args,
+                dryrun=False,
+            )
+
+    # Install a new software
+    else:
+        cli.install(
             args.install_recipe,
             force=args.force,
             container_image=args.container_image,
+            keep_path=args.keep_path,
         )
+        if cli.settings.default_view and not args.no_view:
+            cli.view_install(
+                cli.settings.default_view,
+                args.install_recipe,
+                force=args.force,
+                container_image=args.container_image,
+            )
+
+
+def upgrade(name, cli, args, dryrun=False):
```

Comment:
This (to me) in how  it's designed here looks like it better belongs in an associated helper script. It's not a core part of shpc, it's written as a kind of helper function that uses the client.

Reply:
Hi @vsoch, thank you for the response. Could you please shed more light on this? I assume you're saying `upgrade()` should be defined in a different script rather than in install.py?

Reply:
I still don't think I'm convinced this belongs in core shpc. The implementation here reflects that - it's a big block of code that uses shpc and would better serve as a supplementary helper script.

Reply:
Oh, I'm not disagreeing with your suggestion, I just wanted to understand your thought process more in order to know the next steps. 
From your suggestion, where should the upgrade function be placed, just so we can all be on the same page @vsoch

Reply:
Likely a new subdirectory of https://github.com/singularityhub/singularity-hpc/tree/main/example

Reply:
Great thanks, will work on that.

Should I also add `def get_latest_version(name, config)` there? @vsoch

Reply:
Hi @vsoch . If the functionality moves to a separate script outside of `shpc install`, then no need to name the option "install --upgrade", right ? We can change the name back to "shpc-upgrade.py".

@Ausbeth : compared to where things were in https://github.com/sanger-tol/singularity-hpc/pull/3, you would need to:

- move `shpc/client/upgrade.py` to `example/shpc-upgrade.py`, and embed

   1. the argparse stuff from [`shpc/client/__init__.py`](https://github.com/sanger-tol/singularity-hpc/pull/3/files#diff-850edb9d0d274d69d03fb1ae0b0292baf25d1d85d67629ab87659a0f2ac3f6e4)
   2. the help text from [`shpc/client/help.py`](https://github.com/sanger-tol/singularity-hpc/pull/3/files#diff-9de279ab4e05f61f6ed8fe0964dcbac71dd1bda2691b4aa93e7a59a06bcdf49a)
- move the tests as well

Probably easier to start from 081dc2de017f3e48d405d6edaa40aca54c4c6613 rather than the current state of the branch